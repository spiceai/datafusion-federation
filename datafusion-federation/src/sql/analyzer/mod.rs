mod collect_known_rewrites;
mod rewrite_table_scan;

pub use collect_known_rewrites::collect_known_rewrites;
pub use rewrite_table_scan::*;
