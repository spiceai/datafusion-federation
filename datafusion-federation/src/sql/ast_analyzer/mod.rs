mod rewrite_multi_part;
mod table_arg;

use datafusion::{common::error::Result, sql::sqlparser::ast};
pub use rewrite_multi_part::RewriteMultiTableReference;
pub use table_arg::TableArgReplace;

pub type AstAnalyzerRule = Box<dyn FnMut(ast::Statement) -> Result<ast::Statement>>;

#[derive(Default)]
pub struct AstAnalyzer {
    rules: Vec<AstAnalyzerRule>,
}

impl AstAnalyzer {
    pub fn new(rules: Vec<AstAnalyzerRule>) -> Self {
        Self { rules }
    }

    pub fn add_rule(&mut self, rule: AstAnalyzerRule) {
        self.rules.push(rule)
    }

    pub fn analyze(&mut self, mut statement: ast::Statement) -> Result<ast::Statement> {
        for rule in self.rules.iter_mut() {
            statement = rule(statement)?;
        }

        Ok(statement)
    }
}
