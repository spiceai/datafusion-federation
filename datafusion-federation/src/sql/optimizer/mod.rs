//! A collection of optimizer rules that benefit SQL federation.

mod push_down_filter;
pub use push_down_filter::PushDownFilterFederation;

mod optimize_projections;
pub use optimize_projections::OptimizeProjectionsFederation;
