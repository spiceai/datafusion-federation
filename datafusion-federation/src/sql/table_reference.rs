use std::sync::Arc;

use datafusion::{
    error::DataFusionError,
    sql::{
        sqlparser::{
            self,
            ast::{FunctionArg, ObjectName, ObjectNamePart},
            dialect::{Dialect, GenericDialect},
            tokenizer::Token,
        },
        TableReference,
    },
};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct MultiTableReference {
    parts: Vec<Arc<str>>,
}

impl MultiTableReference {
    pub fn new(parts: Vec<Arc<str>>) -> Self {
        assert!(parts.len() > 3, "Use TableReference, not MultiTableReference, for table references with less than 3 parts.");
        Self { parts }
    }

    pub fn parts(&self) -> &Vec<Arc<str>> {
        &self.parts
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum MultiPartTableReference {
    TableReference(TableReference),
    Multi(MultiTableReference),
}

impl MultiPartTableReference {
    /// Convert a `MultiPartTableReference` to a quoted string.
    ///
    /// Example:
    ///
    /// ```ignore
    /// let parts = vec![Arc::from("a"), Arc::from("b"), Arc::from("c"), Arc::from("d")];
    /// let multi_part_table_reference = MultiPartTableReference::encode_multi_part_table_reference(&parts);
    /// assert_eq!(multi_part_table_reference.to_quoted_string(), r#""a"."b"."c"."d""#);
    /// ```
    #[must_use]
    pub fn to_quoted_string(&self) -> String {
        match self {
            MultiPartTableReference::TableReference(table_reference) => match table_reference {
                // The `TableReference` will sometimes not quote the table name, even if we ask it to because it detects that it would be safe (within DataFusion).
                // Unfortunately, some systems have reserved keywords that will error if we don't quote them.
                // Err on the safe side and always quote the table name.
                TableReference::Bare { table } => quote_identifier(table),
                TableReference::Partial { schema, table } => {
                    format!("{}.{}", quote_identifier(schema), quote_identifier(table))
                }
                TableReference::Full {
                    catalog,
                    schema,
                    table,
                } => format!(
                    "{}.{}.{}",
                    quote_identifier(catalog),
                    quote_identifier(schema),
                    quote_identifier(table)
                ),
            },
            MultiPartTableReference::Multi(parts) => parts
                .iter()
                .map(|p| quote_identifier(p))
                .collect::<Vec<_>>()
                .join("."),
        }
    }

    /// Compare with another [`MultiPartTableReference`] as if both are resolved.
    /// This allows comparing across variants. If a field is not present
    /// in both variants being compared then it is ignored in the comparison.
    ///
    /// e.g. this allows a [`TableReference::Bare`] to be considered equal to a
    /// fully qualified [`TableReference::Full`] if the table names match.
    pub fn resolved_eq(&self, other: &Self) -> bool {
        match self {
            MultiPartTableReference::TableReference(table_reference) => match other {
                MultiPartTableReference::TableReference(other_table_reference) => {
                    table_reference.resolved_eq(other_table_reference)
                }
                MultiPartTableReference::Multi(_) => false,
            },
            MultiPartTableReference::Multi(parts) => match other {
                MultiPartTableReference::Multi(other_parts) => {
                    parts.iter().zip(other_parts.iter()).all(|(a, b)| a == b)
                }
                MultiPartTableReference::TableReference(_) => false,
            },
        }
    }

    pub fn table(&self) -> &str {
        match self {
            MultiPartTableReference::TableReference(table_reference) => table_reference.table(),
            MultiPartTableReference::Multi(parts) => parts.last().unwrap(),
        }
    }
}

impl std::fmt::Display for MultiPartTableReference {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MultiPartTableReference::TableReference(table_reference) => {
                write!(f, "{table_reference}")
            }
            MultiPartTableReference::Multi(parts) => {
                write!(f, "{}", parts.join("."))
            }
        }
    }
}

impl std::ops::Deref for MultiTableReference {
    type Target = Vec<Arc<str>>;

    fn deref(&self) -> &Self::Target {
        &self.parts
    }
}

impl From<ObjectName> for MultiPartTableReference {
    fn from(name: ObjectName) -> Self {
        let mut parts = name.0;
        match parts.len() {
            1 => MultiPartTableReference::TableReference(TableReference::Bare {
                table: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
            }),
            2 => MultiPartTableReference::TableReference(TableReference::Partial {
                schema: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
                table: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
            }),
            3 => MultiPartTableReference::TableReference(TableReference::Full {
                catalog: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
                schema: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
                table: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
            }),
            _ => MultiPartTableReference::Multi(MultiTableReference {
                parts: parts
                    .into_iter()
                    .map(|p| match p {
                        ObjectNamePart::Identifier(ident) => ident.value.into(),
                    })
                    .collect(),
            }),
        }
    }
}

impl From<&ObjectName> for MultiPartTableReference {
    fn from(name: &ObjectName) -> Self {
        name.clone().into()
    }
}

impl From<&RemoteTableRef> for MultiPartTableReference {
    fn from(table_ref: &RemoteTableRef) -> Self {
        table_ref.table_ref().clone()
    }
}

impl From<TableReference> for MultiPartTableReference {
    fn from(table_ref: TableReference) -> Self {
        MultiPartTableReference::TableReference(table_ref)
    }
}

impl From<&TableReference> for MultiPartTableReference {
    fn from(table_ref: &TableReference) -> Self {
        MultiPartTableReference::TableReference(table_ref.clone())
    }
}

/// A multipart identifier to a remote table, view or parameterized view.
///
/// RemoteTableRef can be created by parsing from a string representing a table object with optional
/// ```rust
/// use datafusion_federation::sql::RemoteTableRef;
/// use datafusion::sql::sqlparser::dialect::PostgreSqlDialect;
///
/// RemoteTableRef::try_from("myschema.table");
/// RemoteTableRef::try_from(r#"myschema."Table""#);
/// RemoteTableRef::try_from("myschema.view('obj')");
///
/// RemoteTableRef::parse_with_dialect("myschema.view(name = 'obj')", &PostgreSqlDialect {});
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct RemoteTableRef {
    pub table_ref: MultiPartTableReference,
    pub args: Option<Arc<[FunctionArg]>>,
}

impl RemoteTableRef {
    /// Get quoted_string representation for the table it is referencing, this is same as calling to_quoted_string on the inner table reference.
    pub fn to_quoted_string(&self) -> String {
        self.table_ref.to_quoted_string()
    }

    /// Create new using general purpose dialect. Prefer [`Self::parse_with_dialect`] if the dialect is known beforehand
    pub fn parse_with_default_dialect(s: &str) -> Result<Self, DataFusionError> {
        Self::parse_with_dialect(s, &GenericDialect {})
    }

    /// Create new using a specific instance of dialect.
    pub fn parse_with_dialect(s: &str, dialect: &dyn Dialect) -> Result<Self, DataFusionError> {
        let mut parser = sqlparser::parser::Parser::new(dialect).try_with_sql(s)?;
        let name = parser.parse_object_name(true)?;
        let args = if parser.consume_token(&Token::LParen) {
            parser.parse_optional_args()?
        } else {
            vec![]
        };

        let mut parts = name.0;
        let table_ref = match parts.len() {
            1 => MultiPartTableReference::TableReference(TableReference::Bare {
                table: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
            }),
            2 => MultiPartTableReference::TableReference(TableReference::Partial {
                schema: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
                table: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
            }),
            3 => MultiPartTableReference::TableReference(TableReference::Full {
                catalog: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
                schema: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
                table: match parts.remove(0) {
                    ObjectNamePart::Identifier(ident) => ident.value.into(),
                },
            }),
            _ => MultiPartTableReference::Multi(MultiTableReference {
                parts: parts
                    .into_iter()
                    .map(|p| match p {
                        ObjectNamePart::Identifier(ident) => ident.value.into(),
                    })
                    .collect(),
            }),
        };

        if !args.is_empty() {
            Ok(RemoteTableRef {
                table_ref,
                args: Some(args.into()),
            })
        } else {
            Ok(RemoteTableRef {
                table_ref,
                args: None,
            })
        }
    }

    pub fn table_ref(&self) -> &MultiPartTableReference {
        &self.table_ref
    }

    pub fn args(&self) -> Option<&[FunctionArg]> {
        self.args.as_deref()
    }
}

impl From<TableReference> for RemoteTableRef {
    fn from(table_ref: TableReference) -> Self {
        RemoteTableRef {
            table_ref: MultiPartTableReference::TableReference(table_ref),
            args: None,
        }
    }
}

impl TryFrom<MultiPartTableReference> for TableReference {
    type Error = DataFusionError;

    fn try_from(value: MultiPartTableReference) -> Result<Self, Self::Error> {
        match value {
            MultiPartTableReference::TableReference(table_reference) => Ok(table_reference),
            MultiPartTableReference::Multi(_) => Err(DataFusionError::External(
                "MultiPartTableReference cannot be converted to TableReference".into(),
            )),
        }
    }
}

impl TryFrom<&MultiPartTableReference> for TableReference {
    type Error = DataFusionError;

    fn try_from(value: &MultiPartTableReference) -> Result<Self, Self::Error> {
        match value {
            MultiPartTableReference::TableReference(table_reference) => Ok(table_reference.clone()),
            MultiPartTableReference::Multi(_) => Err(DataFusionError::External(
                "MultiPartTableReference cannot be converted to TableReference".into(),
            )),
        }
    }
}

impl From<MultiPartTableReference> for RemoteTableRef {
    fn from(value: MultiPartTableReference) -> Self {
        RemoteTableRef {
            table_ref: value,
            args: None,
        }
    }
}

impl From<(TableReference, Vec<FunctionArg>)> for RemoteTableRef {
    fn from((table_ref, args): (TableReference, Vec<FunctionArg>)) -> Self {
        RemoteTableRef {
            table_ref: MultiPartTableReference::TableReference(table_ref),
            args: Some(args.into()),
        }
    }
}

impl TryFrom<&str> for RemoteTableRef {
    type Error = DataFusionError;
    fn try_from(s: &str) -> Result<Self, Self::Error> {
        Self::parse_with_default_dialect(s)
    }
}

impl TryFrom<String> for RemoteTableRef {
    type Error = DataFusionError;
    fn try_from(s: String) -> Result<Self, Self::Error> {
        Self::parse_with_default_dialect(&s)
    }
}

impl TryFrom<&String> for RemoteTableRef {
    type Error = DataFusionError;
    fn try_from(s: &String) -> Result<Self, Self::Error> {
        Self::parse_with_default_dialect(s)
    }
}

/// Wraps identifier string in double quotes, escaping any double quotes in
/// the identifier by replacing it with two double quotes
///
/// e.g. identifier `tab.le"name` becomes `"tab.le""name"`
#[must_use]
pub fn quote_identifier(s: &str) -> String {
    format!("\"{}\"", s.replace('"', "\"\""))
}

#[cfg(test)]
mod tests {
    use super::*;
    use datafusion::sql::sqlparser::{ast::ValueWithSpan, tokenizer::Span};
    use sqlparser::{
        ast::{self, Expr, FunctionArgOperator, Ident, Value},
        dialect,
    };

    #[test]
    fn bare_table_reference() {
        let table_ref = RemoteTableRef::parse_with_default_dialect("table").unwrap();
        let expected = RemoteTableRef::from(TableReference::bare("table"));
        assert_eq!(table_ref, expected);

        let table_ref = RemoteTableRef::parse_with_default_dialect("Table").unwrap();
        let expected = RemoteTableRef::from(TableReference::bare("Table"));
        assert_eq!(table_ref, expected);
    }

    #[test]
    fn bare_table_reference_with_args() {
        let table_ref = RemoteTableRef::parse_with_default_dialect("table(1, 2)").unwrap();
        let expected = RemoteTableRef::from((
            TableReference::bare("table"),
            vec![
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("1".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("2".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
            ],
        ));
        assert_eq!(table_ref, expected);

        let table_ref = RemoteTableRef::parse_with_default_dialect("Table(1, 2)").unwrap();
        let expected = RemoteTableRef::from((
            TableReference::bare("Table"),
            vec![
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("1".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("2".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
            ],
        ));
        assert_eq!(table_ref, expected);
    }

    #[test]
    fn bare_table_reference_with_args_and_whitespace() {
        let table_ref = RemoteTableRef::parse_with_default_dialect("table (1, 2)").unwrap();
        let expected = RemoteTableRef::from((
            TableReference::bare("table"),
            vec![
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("1".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("2".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
            ],
        ));
        assert_eq!(table_ref, expected);

        let table_ref = RemoteTableRef::parse_with_default_dialect("Table (1, 2)").unwrap();
        let expected = RemoteTableRef::from((
            TableReference::bare("Table"),
            vec![
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("1".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("2".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
            ],
        ));
        assert_eq!(table_ref, expected);
    }

    #[test]
    fn multi_table_reference_with_no_args() {
        let table_ref = RemoteTableRef::parse_with_default_dialect("schema.table").unwrap();
        let expected = RemoteTableRef::from(TableReference::partial("schema", "table"));
        assert_eq!(table_ref, expected);

        let table_ref = RemoteTableRef::parse_with_default_dialect("schema.Table").unwrap();
        let expected = RemoteTableRef::from(TableReference::partial("schema", "Table"));
        assert_eq!(table_ref, expected);
    }

    #[test]
    fn multi_table_reference_with_args() {
        let table_ref = RemoteTableRef::parse_with_default_dialect("schema.table(1, 2)").unwrap();
        let expected = RemoteTableRef::from((
            TableReference::partial("schema", "table"),
            vec![
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("1".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("2".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
            ],
        ));
        assert_eq!(table_ref, expected);

        let table_ref = RemoteTableRef::parse_with_default_dialect("schema.Table(1, 2)").unwrap();
        let expected = RemoteTableRef::from((
            TableReference::partial("schema", "Table"),
            vec![
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("1".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("2".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
            ],
        ));
        assert_eq!(table_ref, expected);
    }

    #[test]
    fn multi_table_reference_with_args_and_whitespace() {
        let table_ref = RemoteTableRef::parse_with_default_dialect("schema.table (1, 2)").unwrap();
        let expected = RemoteTableRef::from((
            TableReference::partial("schema", "table"),
            vec![
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("1".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
                FunctionArg::Unnamed(
                    Expr::Value(ValueWithSpan {
                        value: Value::Number("2".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                ),
            ],
        ));
        assert_eq!(table_ref, expected);
    }

    #[test]
    fn bare_reference_with_named_args() {
        let table_ref = RemoteTableRef::parse_with_dialect(
            "Table (user_id => 1, age => 2)",
            &dialect::PostgreSqlDialect {},
        )
        .unwrap();
        let expected = RemoteTableRef::from((
            TableReference::bare("Table"),
            vec![
                FunctionArg::ExprNamed {
                    name: ast::Expr::Identifier(Ident::new("user_id")),
                    arg: Expr::Value(ValueWithSpan {
                        value: Value::Number("1".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                    operator: FunctionArgOperator::RightArrow,
                },
                FunctionArg::ExprNamed {
                    name: ast::Expr::Identifier(Ident::new("age")),
                    arg: Expr::Value(ValueWithSpan {
                        value: Value::Number("2".to_string(), false),
                        span: Span::empty(),
                    })
                    .into(),
                    operator: FunctionArgOperator::RightArrow,
                },
            ],
        ));
        assert_eq!(table_ref, expected);
    }
}
