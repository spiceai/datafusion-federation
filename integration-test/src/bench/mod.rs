use crate::validation::validate_tpch_query;
use anyhow::Result;
use datafusion::arrow::record_batch::RecordBatch;
use std::sync::Arc;

pub trait Benchmark {
    /// Name of the benchmark (e.g., "tpch", "tpcds")
    fn name(&self) -> &str;

    /// Database file name (e.g., "tpch_sf1.db", "tpcds_sf1.db")
    fn db_file_name(&self) -> String;

    /// Get all table names for this benchmark
    fn table_names(&self) -> Vec<String>;

    /// Get all queries for this benchmark
    fn queries(&self) -> Vec<Query>;

    fn validate(&self, query: &Query, batches: &[RecordBatch]) -> Result<()>;
}

pub struct Query {
    pub name: Arc<str>,
    pub sql: Arc<str>,
}

macro_rules! generate_tpch_queries {
     ( $( $i:tt ),* ) => {
         vec![
             $(
                 Query {
                     name: concat!("tpch_", stringify!($i)).into(),
                     sql: include_str!(concat!("../../queries/tpch/", stringify!($i), ".sql")).into(),
                 }
             ),*
         ]
     }
}

pub struct TpchBenchmark;

impl Benchmark for TpchBenchmark {
    fn name(&self) -> &str {
        "tpch"
    }

    fn db_file_name(&self) -> String {
        "tpch_sf1.db".to_string()
    }

    fn table_names(&self) -> Vec<String> {
        vec![
            "customer".to_string(),
            "orders".to_string(),
            "lineitem".to_string(),
            "part".to_string(),
            "supplier".to_string(),
            "partsupp".to_string(),
            "nation".to_string(),
            "region".to_string(),
        ]
    }

    fn queries(&self) -> Vec<Query> {
        generate_tpch_queries!(
            q1, q2, q3, q4, q5, q6, q7, q8, q9, q10, q11, q12, q13, q14, q16, q17, q18, q19, q20,
            q21, q22
        )
    }

    fn validate(&self, query: &Query, batches: &[RecordBatch]) -> Result<()> {
        validate_tpch_query(query, batches)
    }
}
