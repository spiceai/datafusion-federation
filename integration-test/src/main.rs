mod bench;
mod validation;

use anyhow::{anyhow, Result};
use bench::{Benchmark, Query};
use std::path::PathBuf;
use std::sync::Arc;

use datafusion::arrow::util::pretty::pretty_format_batches;
use datafusion::sql::TableReference;
use datafusion_federation::{FederatedQueryPlanner, FederationAnalyzerRule};
use datafusion_table_providers::{
    duckdb::DuckDBTableFactory, sql::db_connection_pool::duckdbpool::DuckDbConnectionPool,
};
use duckdb::AccessMode;

use datafusion::{
    execution::{context::SessionContext, session_state::SessionStateBuilder},
    optimizer::{
        analyzer::{
            expand_wildcard_rule::ExpandWildcardRule, inline_table_scan::InlineTableScan,
            resolve_grouping_function::ResolveGroupingFunction, type_coercion::TypeCoercion,
        },
        AnalyzerRule,
    },
};

pub fn get_analyzer_rules() -> Vec<Arc<dyn AnalyzerRule + Send + Sync>> {
    vec![
        Arc::new(InlineTableScan::new()),
        Arc::new(ExpandWildcardRule::new()),
        Arc::new(FederationAnalyzerRule::new()),
        Arc::new(ResolveGroupingFunction::new()),
        Arc::new(TypeCoercion::new()),
    ]
}

fn get_duckdb_table_factory(db_name: String) -> Result<DuckDBTableFactory> {
    let db_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .ok_or_else(|| anyhow!("Failed to get parent directory"))?
        .join(db_name);

    let duckdb_pool = Arc::new(
        DuckDbConnectionPool::new_file(
            db_path
                .to_str()
                .ok_or_else(|| anyhow!("Failed to convert db_path to string"))?,
            &AccessMode::ReadOnly,
        )
        .map_err(|e| anyhow!("Failed to create DuckDB connection pool: {}", e))?,
    );

    Ok(DuckDBTableFactory::new(duckdb_pool))
}

#[tokio::main]
async fn main() {
    let benchmark: Arc<dyn Benchmark> = Arc::new(bench::TpchBenchmark);

    let duckdb_table_factory = get_duckdb_table_factory(benchmark.db_file_name())
        .expect("unable to create DuckDB connection pool");

    let state = SessionStateBuilder::new()
        .with_query_planner(Arc::new(FederatedQueryPlanner::new()))
        .with_analyzer_rules(get_analyzer_rules())
        .with_default_features()
        .build();
    let ctx = SessionContext::new_with_state(state);

    let _ = register_federated_duckdb_tables(&ctx, benchmark.table_names(), &duckdb_table_factory)
        .await;

    let test_queries = benchmark.queries();

    for query in test_queries {
        run_test_query(&ctx, Arc::clone(&benchmark), &query)
            .await
            .unwrap_or_else(|err| {
                panic!(
                    "Failed to run {} query {}: {}",
                    benchmark.name(),
                    query.name,
                    err
                )
            });
    }
}

async fn run_test_query(
    ctx: &SessionContext,
    benchmark: Arc<dyn Benchmark>,
    query: &Query,
) -> Result<()> {
    let df = ctx.sql(&query.sql).await?;

    let plan = df.clone().explain(false, false)?.collect().await?;
    let plan_display = pretty_format_batches(&plan)?;

    insta::with_settings!({
        description => format!("Federated Query Explain"),
        snapshot_path => "../snapshots/explain",
        filters => vec![
            (r"compute_context=.*/([^/]+\.db)", "compute_context=$1")
        ],
    }, {
        insta::assert_snapshot!(format!("{}_{}_explain", benchmark.name(), query.name), plan_display);
    });

    let result = df.collect().await?;
    benchmark.validate(query, &result)?;

    Ok(())
}

async fn register_federated_duckdb_tables(
    ctx: &SessionContext,
    table_names: Vec<String>,
    duckdb_table_factory: &DuckDBTableFactory,
) -> Result<()> {
    for table_name in table_names {
        ctx.register_table(
            &table_name,
            duckdb_table_factory
                .table_provider(TableReference::bare(table_name.as_str()))
                .await
                .map_err(|e| anyhow!("Failed to register duckdb table: {}", e))?,
        )?;
    }

    Ok(())
}
