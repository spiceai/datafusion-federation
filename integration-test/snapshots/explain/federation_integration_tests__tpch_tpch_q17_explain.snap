---
source: integration-test/src/main.rs
description: Federated Query Explain
expression: plan_display
---
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
|               |  Projection: sum(lineitem.l_extendedprice) / Float64(7) AS avg_yearly                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|               |   Aggregate: groupBy=[[]], aggr=[[sum(lineitem.l_extendedprice)]]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
|               |     Inner Join:  Filter: part.p_partkey = lineitem.l_partkey                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               |       Filter: lineitem.l_quantity < (<subquery>)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
|               |         Subquery:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
|               |           Projection: Float64(0.2) * avg(lineitem.l_quantity)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|               |             Aggregate: groupBy=[[]], aggr=[[avg(lineitem.l_quantity)]]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|               |               Filter: lineitem.l_partkey = outer_ref(part.p_partkey)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
|               |                 TableScan: lineitem                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
|               |         TableScan: lineitem projection=[l_partkey, l_quantity, l_extendedprice]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
|               |       TableScan: part projection=[p_partkey], full_filters=[part.p_brand = Utf8("Brand#23"), part.p_container = Utf8("MED BOX")]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| physical_plan | SchemaCastScanExec                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
|               |   VirtualExecutionPlan name=duckdb compute_context=tpch_sf1.db sql=SELECT (sum(lineitem.l_extendedprice) / 7.0) AS avg_yearly FROM lineitem JOIN "part" ON (("part".p_partkey = lineitem.l_partkey) AND (((lineitem.l_quantity < (SELECT (0.2 * avg(lineitem.l_quantity)) FROM lineitem WHERE (lineitem.l_partkey = "part".p_partkey))) AND ("part".p_brand = 'Brand#23')) AND ("part".p_container = 'MED BOX'))) rewritten_sql=SELECT (sum("lineitem"."l_extendedprice") // 7.0) AS "avg_yearly" FROM "lineitem" JOIN "part" ON (("part"."p_partkey" = "lineitem"."l_partkey") AND ((("lineitem"."l_quantity" < (SELECT (0.2 * avg("lineitem"."l_quantity")) FROM "lineitem" WHERE ("lineitem"."l_partkey" = "part"."p_partkey"))) AND ("part"."p_brand" = 'Brand#23')) AND ("part"."p_container" = 'MED BOX'))) |
|               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
